### PR TITLE
Upgrading dependency versions as well as GitHub Action versions

### PR DESCRIPTION
**Pull request by cflynn3**

### CODE REVIEW

**Comment by cflynn3** [comment]
I've upgraded various dependency versions, GitHub actions, as well as fixed a couple of newly discovered MyPy issues. Unfortunately we can't currently upgrade to allow Python 3.11 at this time as [numba still doesn't supprort it](https://github.com/numba/numba/issues/8304), but when they do allow upgrades of the Python version it should be simpler to do now.

File: pyproject.toml [hunk 1]
```diff
@@ -39,28 +39,29 @@ include = ["data/*.json"]
...
-python = ">= 3.8, < 3.11"
-numpy = "^1.22"
-pandas = "^1.1"
-h5py = "^3.2"
-lxml = "^4.6"
-lasio = "^0.29"
-scipy = "^1.7.3"
-numba = ">=0.52, < 1.0"
-joblib = "^1.1.0"
+python = ">= 3.8.1, < 3.12"
```
Comment (andy-beer) [inline pyproject.toml:42]: is the < 3.12 okay here, given the current numba limitation?

**Review by andy-beer — COMMENTED** [review]
Only one small query from me (see comment). Otherwise good to merge. Thanks Chris.

File: pyproject.toml (hunk 1 above)
Comment (cflynn3) [reply pyproject.toml]: Yep good catch, I've changed it back for now.

**Pull request merged**